File: src/playback/chord_engine.py
"""
Roblox Piano Player - Chord & Mixed Key Playback Engine
"""
import time
from enum import Enum
from typing import List, Dict, Tuple, Optional, Callable
from src.music.events import NoteEvent
from src.piano.mapper import RobloxPianoMapper
from src.piano.profile import KeyMapping
from src.playback.key_state_manager import KeyStateManager


class ConflictPolicy(Enum):
    MICRO_ARPEGGIO = "MICRO_ARPEGGIO"
    SKIP_CONFLICTED = "SKIP_CONFLICTED"
    WARN_ONLY = "WARN_ONLY"


class ChordEngine:
    """
    Handles mixed Shift/Unshifted chords and resolves same physical key conflicts
    (e.g., q + Q sharing the same 'Q' physical key).
    """

    def __init__(
        self,
        key_state: KeyStateManager,
        mapper: RobloxPianoMapper,
        conflict_policy: ConflictPolicy = ConflictPolicy.MICRO_ARPEGGIO,
        conflict_delay_ms: float = 8.0,
        default_hold_duration_ms: float = 30.0,
        on_log: Optional[Callable[[str], None]] = None
    ):
        self.key_state: KeyStateManager = key_state
        self.mapper: RobloxPianoMapper = mapper
        self.conflict_policy: ConflictPolicy = conflict_policy
        self.conflict_delay_ms: float = conflict_delay_ms
        self.default_hold_duration_ms: float = default_hold_duration_ms
        self.on_log: Optional[Callable[[str], None]] = on_log

    def _log(self, message: str) -> None:
        if self.on_log:
            self.on_log(message)

    def play_chord_notes(
        self,
        notes: List[NoteEvent],
        hold_duration_ms: Optional[float] = None
    ) -> None:
        """
        Executes a chord of NoteEvents with proper modifier isolation and conflict resolution.
        """
        if not notes:
            return

        hold_ms = hold_duration_ms or self.default_hold_duration_ms
        hold_sec = max(0.01, hold_ms / 1000.0)
        delay_sec = self.conflict_delay_ms / 1000.0

        # 1. Map notes to physical keys
        mapped_keys: List[Tuple[NoteEvent, KeyMapping]] = []
        for n in notes:
            km = self.mapper.map_pitch(n.pitch)
            if km:
                mapped_keys.append((n, km))

        if not mapped_keys:
            return

        # 2. Check for same physical key conflicts
        physical_key_map: Dict[str, List[KeyMapping]] = {}
        for _, km in mapped_keys:
            pk = km.physical_key.lower()
            physical_key_map.setdefault(pk, []).append(km)

        conflicted_keys = {pk: kms for pk, kms in physical_key_map.items() if len(kms) > 1}

        if conflicted_keys and self.conflict_policy == ConflictPolicy.SKIP_CONFLICTED:
            # Keep only the highest pitch note for each physical key
            filtered_mapped = []
            seen_pk = set()
            for n, km in sorted(mapped_keys, key=lambda item: item[0].pitch, reverse=True):
                pk = km.physical_key.lower()
                if pk not in seen_pk:
                    seen_pk.add(pk)
                    filtered_mapped.append((n, km))
            mapped_keys = filtered_mapped

        # 3. Group by modifier sets (Execution Plan)
        # We must serialize different modifier groups to prevent modifier bleeding.
        # e.g., if we press Shift for black keys, we don't want it to affect Normal keys.
        modifier_groups: Dict[frozenset, List[KeyMapping]] = {}
        for _, km in mapped_keys:
            modifier_groups.setdefault(km.modifiers, []).append(km)

        # Sort groups: empty modifiers first, then others, to maintain a consistent execution order
        sorted_mod_groups = sorted(modifier_groups.items(), key=lambda x: (len(x[0]), list(x[0])))

        is_multi_group = len(sorted_mod_groups) > 1 or conflicted_keys

        if is_multi_group:
            # Micro-Arpeggio execution plan
            for i, (mods, kms) in enumerate(sorted_mod_groups):
                # Set modifiers for this group
                for mod in mods:
                    self.key_state.set_modifier(mod, True)
                if mods:
                    time.sleep(0.002)

                # Press keys
                for km in kms:
                    self.key_state.press_physical_key(km.physical_key)
                
                # Hold for micro duration if not the last group, else full duration
                if i < len(sorted_mod_groups) - 1:
                    time.sleep(delay_sec)
                else:
                    time.sleep(hold_sec)

                # Release keys
                for km in kms:
                    self.key_state.release_physical_key(km.physical_key)
                
                # Release modifiers
                for mod in mods:
                    self.key_state.set_modifier(mod, False)
                if mods:
                    time.sleep(0.002)
        else:
            # Single modifier group, standard execution
            mods, kms = sorted_mod_groups[0]
            
            for mod in mods:
                self.key_state.set_modifier(mod, True)
            if mods:
                time.sleep(0.002)

            for km in kms:
                self.key_state.press_physical_key(km.physical_key)

            time.sleep(hold_sec)

            for km in kms:
                self.key_state.release_physical_key(km.physical_key)

            for mod in mods:
                self.key_state.set_modifier(mod, False)

File: v2/src/RobloxPiano.Playback.Windows/Playback/ChordEngine.cs
using System.Diagnostics;
using RobloxPiano.Core.Music;
using RobloxPiano.Core.Piano;

namespace RobloxPiano.Playback.Windows.Playback;

public class ChordEngine
{
    private readonly KeyStateManager _keyState;
    private readonly RobloxPianoMapper _mapper;
    private readonly ConflictPolicy _conflictPolicy;
    private readonly double _conflictDelayMs;
    private readonly double _defaultHoldDurationMs;
    private readonly double _modifierSettleMs;
    private readonly double _randomizationMs;

    public KeyStateManager KeyState => _keyState;
    public RobloxPianoMapper Mapper => _mapper;
    public ConflictPolicy ConflictPolicy => _conflictPolicy;

    public ChordEngine(
        KeyStateManager keyState,
        RobloxPianoMapper mapper,
        ConflictPolicy conflictPolicy = ConflictPolicy.MicroArpeggio,
        double conflictDelayMs = 8.0,
        double defaultHoldDurationMs = 30.0,
        double modifierSettleMs = 2.0,
        double randomizationMs = 0.0)
    {
        _keyState = keyState;
        _mapper = mapper;
        _conflictPolicy = conflictPolicy;
        _conflictDelayMs = conflictDelayMs;
        _defaultHoldDurationMs = defaultHoldDurationMs;
        _modifierSettleMs = modifierSettleMs;
        _randomizationMs = randomizationMs;
    }

    public ChordPlaybackResult PlayChordNotes(
        IReadOnlyList<NoteEvent> notes,
        double? holdDurationMs = null,
        int transpose = 0,
        CancellationToken ct = default)
    {
        ct.ThrowIfCancellationRequested();

        if (notes == null || notes.Count == 0)
        {
            return new ChordPlaybackResult(notes?.Count ?? 0, 0, 0, 0, Array.Empty<int>());
        }

        int requestedCount = notes.Count;
        double holdMs = holdDurationMs ?? _defaultHoldDurationMs;
        if (holdMs < 1.0) holdMs = 1.0;

        // 1. Map notes to physical keys with transpose
        var mappedKeys = new List<(NoteEvent Note, KeyMapping Mapping)>();
        int skippedUnmapped = 0;

        foreach (var n in notes)
        {
            int effectivePitch = n.Pitch + transpose;
            var km = _mapper.MapPitch(effectivePitch);
            if (km != null)
            {
                mappedKeys.Add((n, km));
            }
            else
            {
                skippedUnmapped++;
            }
        }

        if (mappedKeys.Count == 0)
        {
            return new ChordPlaybackResult(requestedCount, 0, skippedUnmapped, 0, Array.Empty<int>());
        }

        // 2. Check for same physical key conflicts
        var physicalKeyMap = new Dictionary<string, List<KeyMapping>>(StringComparer.OrdinalIgnoreCase);
        foreach (var (_, km) in mappedKeys)
        {
            var pk = km.PhysicalKey.ToLowerInvariant();
            if (!physicalKeyMap.TryGetValue(pk, out var list))
            {
                list = new List<KeyMapping>();
                physicalKeyMap[pk] = list;
            }
            list.Add(km);
        }

        bool hasPhysicalConflicts = physicalKeyMap.Values.Any(list => list.Count > 1);
        int skippedConflicts = 0;

        if (hasPhysicalConflicts && _conflictPolicy == ConflictPolicy.SkipConflicted)
        {
            // Keep only highest pitch note for each physical key
            var filteredMapped = new List<(NoteEvent Note, KeyMapping Mapping)>();
            var seenPk = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

            foreach (var item in mappedKeys.OrderByDescending(x => x.Note.Pitch))
            {
                var pk = item.Mapping.PhysicalKey.ToLowerInvariant();
                if (seenPk.Add(pk))
                {
                    filteredMapped.Add(item);
                }
                else
                {
                    skippedConflicts++;
                }
            }
            mappedKeys = filteredMapped;
        }

        var playedPitches = mappedKeys.Select(m => m.Note.Pitch + transpose).ToList();

        // 3. Group by modifier sets
        var modifierGroups = new Dictionary<string, (IReadOnlySet<string> Modifiers, List<KeyMapping> Keys)>();

        foreach (var (_, km) in mappedKeys)
        {
            var keyStr = string.Join("+", km.Modifiers.OrderBy(m => m));
            if (!modifierGroups.TryGetValue(keyStr, out var group))
            {
                group = (km.Modifiers, new List<KeyMapping>());
                modifierGroups[keyStr] = group;
            }
            group.Keys.Add(km);
        }

        // Sort groups: empty modifiers first, then by modifier count and names to prevent modifier bleeding
        var sortedGroups = modifierGroups.Values
            .OrderBy(g => g.Modifiers.Count)
            .ThenBy(g => string.Join("+", g.Modifiers.OrderBy(m => m)))
            .ToList();

        bool isMultiGroup = sortedGroups.Count > 1 || hasPhysicalConflicts;

        // Local tracking for guaranteed balanced cleanup in try/finally
        var locallyPressedKeys = new List<string>();
        var locallyPressedModifiers = new List<string>();

        try
        {
            if (isMultiGroup)
            {
                // Micro-Arpeggio execution
                for (int i = 0; i < sortedGroups.Count; i++)
                {
                    ct.ThrowIfCancellationRequested();

                    var (mods, kms) = sortedGroups[i];

                    // Set modifiers for this group
                    foreach (var mod in mods)
                    {
                        ct.ThrowIfCancellationRequested();
                        _keyState.SetModifier(mod, true);
                        locallyPressedModifiers.Add(mod);
                    }
                    if (mods.Count > 0 && _modifierSettleMs > 0)
                    {
                        PreciseDelay(_modifierSettleMs, ct);
                    }

                    // Press physical keys
                    foreach (var km in kms)
                    {
                        ct.ThrowIfCancellationRequested();
                        _keyState.PressPhysicalKey(km.PhysicalKey);
                        locallyPressedKeys.Add(km.PhysicalKey);
                    }

                    // Hold duration: conflict delay for intermediate groups, full duration for the last group
                    double delay = (i < sortedGroups.Count - 1) ? _conflictDelayMs : holdMs;
                    PreciseDelay(delay, ct);

                    // Release physical keys
                    foreach (var km in kms)
                    {
                        _keyState.ReleasePhysicalKey(km.PhysicalKey);
                        locallyPressedKeys.Remove(km.PhysicalKey);
                    }

                    // Release modifiers
                    foreach (var mod in mods)
                    {
                        _keyState.SetModifier(mod, false);
                        locallyPressedModifiers.Remove(mod);
                    }
                    if (mods.Count > 0 && _modifierSettleMs > 0)
                    {
                        PreciseDelay(_modifierSettleMs, ct);
                    }
                }
            }
            else
            {
                // Single modifier group standard execution
                var (mods, kms) = sortedGroups[0];

                foreach (var mod in mods)
                {
                    ct.ThrowIfCancellationRequested();
                    _keyState.SetModifier(mod, true);
                    locallyPressedModifiers.Add(mod);
                }
                if (mods.Count > 0 && _modifierSettleMs > 0)
                {
                    PreciseDelay(_modifierSettleMs, ct);
                }

                foreach (var km in kms)
                {
                    ct.ThrowIfCancellationRequested();
                    _keyState.PressPhysicalKey(km.PhysicalKey);
                    locallyPressedKeys.Add(km.PhysicalKey);
                }

                PreciseDelay(holdMs, ct);

                foreach (var km in kms)
                {
                    _keyState.ReleasePhysicalKey(km.PhysicalKey);
                    locallyPressedKeys.Remove(km.PhysicalKey);
                }

                foreach (var mod in mods)
                {
                    _keyState.SetModifier(mod, false);
                    locallyPressedModifiers.Remove(mod);
                }
            }
        }
        finally
        {
            // Defensive cleanup: always release any remaining keys/modifiers in reverse order
            foreach (var pk in locallyPressedKeys.ToList())
            {
                try
                {
                    _keyState.ReleasePhysicalKey(pk);
                }
                catch { }
            }
            locallyPressedKeys.Clear();

            foreach (var mod in locallyPressedModifiers.ToList())
            {
                try
                {
                    _keyState.SetModifier(mod, false);
                }
                catch { }
            }
            locallyPressedModifiers.Clear();
        }

        return new ChordPlaybackResult(
            requestedCount,
            mappedKeys.Count,
            skippedUnmapped,
            skippedConflicts,
            playedPitches
        );
    }

    private static void PreciseDelay(double milliseconds, CancellationToken ct)
    {
        if (milliseconds <= 0) return;
        ct.ThrowIfCancellationRequested();

        long start = Stopwatch.GetTimestamp();
        long targetTicks = (long)(milliseconds * Stopwatch.Frequency / 1000.0);

        if (milliseconds > 2.0)
        {
            int sleepMs = (int)(milliseconds - 1.5);
            if (sleepMs > 0)
            {
                if (ct.WaitHandle.WaitOne(sleepMs))
                {
                    ct.ThrowIfCancellationRequested();
                }
            }
        }

        while (Stopwatch.GetTimestamp() - start < targetTicks)
        {
            ct.ThrowIfCancellationRequested();
            Thread.Yield();
        }
    }
}
